### PR TITLE
Add backwards compatibility for older versions of Node

### PR DESCRIPTION
**Pull request by PaulAsjes**
## Description
workos-node v6 added support for other JS runtimes, but that meant that the minimum node version required jumped up to v19.

This adds in backwards compatibility to make the SDK usable with at least Node v16. 

## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[ ] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

**Comment by jasonroelofs** [comment]
We should also remove Node 14 from the CI matrix.

**Comment by jasonroelofs** [comment]
Can probably need to tell the linter to ignore these vendored polyfills.

File: src/workos.spec.ts [hunk 1]
```diff
@@ -249,14 +296,35 @@ describe('WorkOS', () => {
     });
 
     describe('when the entity is null', () => {
-      it('sends a null body', async () => {
+      it('sends an empty string body', async () => {
```
Comment (PaulAsjes) [inline src/workos.spec.ts:299]: In fetch environments we can't have the body be null due to how Deno works:

https://github.com/workos/workos-node/pull/1038/files#diff-eb8ece7bf63e57536fd07d4e0ef7369670a03714e326e2a55c1c4a947993fb5dR119-R121

File: src/common/crypto/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export * from './NodeCryptoProvider';
+export * from './SubtleCryptoProvider';
+export * from './CryptoProvider';
```
[the remark this answers is not in the document]
Comment (PaulAsjes) [reply src/common/crypto/index.ts:3]: Already have! Tests are here: https://github.com/workos/workos-node/pull/1038/files#diff-6e2a0e50ff73fa9faf5240a7705aabf4e870e3c3126c877fafc136cc723d346bR227-R270

File: src/common/crypto/CryptoProvider.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Interface encapsulating the various crypto computations used by the library,
+ * allowing pluggable underlying crypto implementations.
+ */
+export class CryptoProvider {
```
[the remark this answers is not in the document]
Comment (PaulAsjes) [reply src/common/crypto/CryptoProvider.ts:5]: Only that I had no idea you could do that! Updated to make this and `HttpClient` abstract classes

File: src/webhooks/webhooks.ts [hunk 4]
```diff
@@ -1,9 +1,22 @@
...
+  SubtleCryptoProvider,
+  NodeCryptoProvider,
+} from '../common/crypto';
 
 export class Webhooks {
-  private encoder = new TextEncoder();
+  private cryptoProvider: CryptoProvider;
+
+  constructor(subtleCrypto?: typeof crypto.subtle) {
+    if (typeof crypto.subtle !== 'undefined') {
```
Comment (alisherry) [inline src/webhooks/webhooks.ts:14]: just double checking, will this be `undefined` as a string?

**Review by alisherry — APPROVED** [review]

File: src/webhooks/webhooks.ts (hunk 4 above)
Comment (PaulAsjes) [reply src/webhooks/webhooks.ts:14]: Yup, because `typeof` always returns the string representation of the type.

**Comment by PaulAsjes** [comment]
Thanks for the review @alisherry! While doing some testing I found some bugs in the Node implementation which should now be fixed. Can you have another look?

File: src/common/crypto/NodeCryptoProvider.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      return false;
+    }
+
+    // Generate a random key for HMAC
+    const key = crypto.randomBytes(32); // Generates a 256-bit key
+    const hmacA = crypto.createHmac('sha256', key).update(bufferA).digest();
+    const hmacB = crypto.createHmac('sha256', key).update(bufferB).digest();
+
+    // Perform a constant time comparison
+    return hmacA.equals(hmacB);
```
Comment (mattgd) [inline src/common/crypto/NodeCryptoProvider.ts:40]: Is `.equals()` on a buffer actually a constant time comparison? I couldn't find any docs that confirm. Or are we able to use `crypto.timingSafeEqual()`?

Comment (PaulAsjes) [reply src/common/crypto/NodeCryptoProvider.ts]: Great catch, turns out GPT lied to me.

**Review by alisherry — APPROVED** [review]

**Pull request merged**